uart/gps_uart/gps_uart.c

```diff
@@ -0,0 +1,250 @@
+        return false;
+    }
+
+    // Ignore all NMEA strings that are not of type GPRMC
+    if (strncmp(nmea_string, "$GPRMC", 6) != 0)
+    {
+        return false;
+    }
+    
+    char *token;
+    char *tokens[12] = {0};
+    int token_count = 0;
+
+    // Tokenize the string
+    token = strtok(nmea_string, ",");
+    while (token != NULL && token_count < 12)
+    {
+        tokens[token_count++] = token;
+
+        token = strtok(NULL, ",");
+    }
+
+    // Check if we have enough tokens and data is valid
+    if (token_count >= 12 && strcmp(tokens[2], "A") == 0)
+    {
+        if (strlen(tokens[3]) > 0)
+        {
+            convert_nmea_to_decimal(tokens[3], &gps_data->latitude);
+            if (tokens[4][0] == 'S')
+                gps_data->latitude = -gps_data->latitude;
+        }
+
+        if (strlen(tokens[5]) > 0)
+        {
+            convert_nmea_to_decimal(tokens[5], &gps_data->longitude);
+            if (tokens[6][0] == 'W')
+                gps_data->longitude = -gps_data->longitude;
+        }
+
+        gps_data->is_valid = true;
+        return true;
+    }
+}
+
+
+/**
+ * @brief Processes incoming GPS data from the UART interface.
+ *
+ * This function reads NMEA sentences from the UART interface, validates them,
+ * and updates the GPSData structure if a valid sentence is received.
+ *
+ * @param gps_data Pointer to the GPSData structure to store the parsed data.
+ */
+void process_gps_data(GPSData *gps_data)
+{
+    char nmea_buffer[MAX_NMEA_LENGTH];
+    int chars_read = 0;
+
+    while (uart_is_readable(UART_ID) && chars_read < MAX_NMEA_LENGTH - 1)
+    {
+        char received_char = uart_getc(UART_ID);
+        nmea_buffer[chars_read] = received_char;
+
+        // NMEA strings are terminated by a newline character
+        if ((int)received_char == 10)
+        {
+            nmea_buffer[chars_read + 1] = '\0';
+
+            if (parse_nmea_gps(nmea_buffer, gps_data))
+            {
+                printf("Valid GPS Data Received\n");
+            }
+
+            chars_read = 0;
+            break;
+        }
+        else
+        {
+            chars_read++;
+        }
+    }
+}
+
+/**
+ * @brief Main function for the GPS driver example.
+ *
+ * This function initializes the UART interface, waits for the GPS module to
+ * cold start, and continuously processes and prints GPS data.
+ */
+int main()
+{
+  stdio_init_all();
+
+  uart_gps_init();
+
+  // recommended cold start waiting time (could vary based on the gps module)
+  sleep_ms(7 * 60 * 1000);
+
+  GPSData gps_data = {0};
+  
+  while (1)
+  {
+    process_gps_data(&gps_data);
+
+    if (gps_data.is_valid)
+    {
+      printf("GPS Location:\n");
+      printf("Latitude: %.6f\n", gps_data.latitude);
+      printf("Longitude: %.6f\n", gps_data.longitude);
+      gps_data.is_valid = false;
+    }
+
+    // sufficient waiting time required between each gps reading and the next one
+    sleep_ms(30 * 1000);
+    tight_loop_contents();
```

Comment:
No need for `tight_loop_contents` here - it's only intended for busy-wait-loops in which nothing else is happening.